doc/source/guide/derivatives_guide.rst

```diff
@@ -0,0 +1,155 @@
+
+* **Derivative**. When we write derivative in ODL code and documentation, we intend the derivative of an `Operator` :math:`A : \mathcal{X} \rightarrow \mathcal{Y}` w.r.t to a disturbance in :math:`x`, i.e a linear approximation of :math:`A(x + h)` for small :math:`h`. The derivative in a point :math:`x` is an `Operator` :math:`A'(x) : \mathcal{X} \rightarrow \mathcal{Y}`.
+
+* **Gradient**. If the operator :math:`A` is a `functional`, i.e. :math:`A : \mathcal{X} \rightarrow \mathbb{R}`, then the gradient is the direction in which :math:`A` increases the most. The gradient in a point :math:`x` is a vector :math:`[\nabla A](x)` in :math:`\mathcal{X}` such that :math:`A'(x)(y) = \langle [\nabla A](x), y \rangle`. The gradient operator is the operator :math:`x \rightarrow [\nabla A](x)`.
+
+* **Hessian**. The hessian in a point :math:`x` is the derivative operator of the gradient operator, i.e. :math:`H(x) = [\nabla A]'(x)`.
+
+* **Spatial Gradient**. The spatial gradient is only defined for spaces :math:`\mathcal{F}(\Omega, \mathbb{F})` whose elements are functions over some domain :math:`\Omega \subset \mathbb{R}^d` taking values in :math:`\mathbb{R}` or :math:`\mathbb{C}`. It can be seen as a vectorized version of the usual gradient, taken in each point in :math:`\Omega`.
+
+Derivative
+##########
+
+The derivative is usually introduced for functions :math:`f: \mathbb{R} \rightarrow \mathbb{R}` via the limit
+
+.. math::
+    f'(x) = \lim_{h \rightarrow 0} \frac{f(x + h) - f(x)}{h}.
+
+Here we say that the derivative of :math:`f` in :math:`x` is :math:`f'(x)`.
+
+This limit makes sense in one dimension, but once we start considering functions in higher dimension we get into trouble.
+Consider :math:`f: \mathbb{R}^n \rightarrow \mathbb{R}^m` -- what would :math:`h` mean in this case?
+An extension is the concept of a directional derivative.
+The derivative of :math:`f` in :math:`x` in *direction* :math:`d` is :math:`f'(x)(d)`:
+
+.. math::
+    f'(x)(d) = \lim_{h \rightarrow 0} \frac{f(x + dh) - f(x)}{h}.
+
+Here we see (as implied by the notation) that :math:`f'(x)` is actually an operator
+
+.. math::
+    f'(x) : \mathbb{R}^n \rightarrow \mathbb{R}^m.
+
+We can rewrite this using the explicit requirement that :math:`f'(x)` is a linear approximation of :math:`f` at :math:`x`, i.e.
+
+.. math::
+   \lim_{\| d \| \rightarrow 0} \frac{\| f(x + d) - f(x) - f'(x)(d) \|}{\| d \|} = 0
+
+This notion naturally extends to an `Operator` :math:`f : \mathcal{X} \rightarrow \mathcal{Y}` between Banach spaces :math:`\mathcal{X}` and :math:`\mathcal{Y}` with norms :math:`\| \cdot \|_\mathcal{X}` and :math:`\| \cdot \|_\mathcal{Y}`, respectively.
+Here :math:`f'(x)` is defined as the linear operator (if it exists) that satisfies
+
+.. math::
+   \lim_{\| d \| \rightarrow 0} \frac{\| f(x + d) - f(x) - f'(x)(d) \|_\mathcal{Y}}{\| d \|_\mathcal{X}} = 0
+
+This definition of the derivative is called the *Fréchet derivative*.
```

Comment:
Perhaps mention that the limit has to be uniform in the directions, which makes it a stronger notion than the Gateaux derivative that only has to exist for each direction.